## Merging the ATO Dataset with the Postcode Data
In this exercise, we will merge the ATO dataset (28 columns) with the Postcode dataset (150 columns) to get a richer dataset with an increased number of columns.

In [1]:
import pandas as pd

In [3]:
ato_url = ('https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter12/Dataset/taxstats2015.csv')

In [5]:
postcode_url = ('https://github.com/PacktWorkshops/'
               'The-Data-Science-Workshop/blob/'
               'master/Chapter12/Dataset/'
               'taxstats2016individual06taxablestatusstate'
               'territorypostcodetaxableincome%20(2).xlsx?'
               'raw=true')

In [6]:
df = pd.read_csv(ato_url)
df.shape

(2473, 28)

In [7]:
df.head()

Postcode  Count taxable income or loss  Average taxable income or loss  \
0      2000                         36185                           47723   
1      2006                            83                           80905   
2      2007                          4769                           46549   
3      2008                          5607                          108816   
4      2009                          9726                           82938   

   Median taxable income or loss  Count salary and wages  \
0                          18213                   31293   
1                          58150                      74   
2                          31474                    4153   
3                          41151                    5008   
4                          50604                    8167   

   Average salary and wages  Median salary and wages  Count net rent  \
0                     38710                    17992            3614   
1                     82733                    67658              21   
2                     47386                    34366             548   
3                     53418                    42892             612   
4                     74068                    55243            1714   

   Average net rent  Median net rent  ...  Median total business income  \
0               558              207  ...                         19684   
1              1042             -807  ...                         42054   
2             -1242             -794  ...                         19960   
3              1818            -1728  ...                         19722   
4             -2412            -1520  ...                         26044   

   Count  total business expenses  Average total business expenses  \
0                            1331                           222191   
1                               4                            48971   
2                             334                           639499   
3                             481                            32173   
4                             666                           222993   

   Median total business expenses  Count net tax  Average net tax  \
0                           20773          17264            27555   
1                           18508             67            28142   
2                            9233           3071            15649   
3                            9385           4061            53976   
4                           11982           7358            32430   

   Median net tax  Count super total accounts balance  \
0            5917                               51394   
1           19359                                 137   
2            9115                                6711   
3           10583                                7676   
4           14835                               12718   

   Average super total accounts balance  Median super total accounts balance  
0                                 68531                                 1110  
1                                 90315                                15544  
2                                 41670                                 5225  
3                                 54245                                10373  
4                                126486                                20494  

[5 rows x 28 columns]

We are interested in learning more about each of these postcodes. Let's make sure they are all unique in this dataset.

In [8]:
# number of unique values for Postcode variable
df['Postcode'].nunique()

2473

We will only load the Individuals Table 6B sheet as this is where the data is located so we need to provide this name to the sheet_name parameter. Also, the header row (containing the name of the variables) in this spreadsheet is located on the third row so we need to specify it to the header parameter.

In [9]:
postcode_df = pd.read_excel(postcode_url, 
                           sheet_name='Individuals Table 6B', 
                           header=2)
postcode_df.shape

(2567, 150)

In [10]:
postcode_df.head()

State/ Territory1 Postcode  Number of individuals\nno.  \
0               ACT     2600                        5581   
1               ACT     2601                        2658   
2               ACT     2602                       19457   
3               ACT     2603                        6478   
4               ACT     2604                        7387   

   Taxable income or loss3 \nno.  Taxable income or loss3 \n$  \
0                           5557                    569612119   
1                           2637                    183288078   
2                          19367                   1379212924   
3                           6439                    796053195   
4                           7341                    637350002   

   Tax on taxable income \nno.  Tax on taxable income \n$  \
0                         4905                  163820305   
1                         1941                   46542933   
2                        16595                  326786223   
3                         5701                  249451322   
4                         6534                  165676071   

   Medicare levy \nno.  Medicare levy \n$  Medicare levy surcharge \nno.  ...  \
0                 4384           10548612                            119  ...   
1                 1660            3203840                             88  ...   
2                14594           24937105                            565  ...   
3                 5176           15002971                            131  ...   
4                 5787           11512325                            248  ...   

   Small business income tax offset \n$  Low income tax offset \nno.  \
0                                125198                         1812   
1                                 30823                          812   
2                                212891                         7623   
3                                149987                         2056   
4                                 73942                         2391   

   Low income tax offset \n$  People with private health insurance \nno.  \
0                     527925                                        4597   
1                     250431                                        1637   
2                    2306744                                       13434   
3                     611022                                        5193   
4                     673244                                        5465   

   Private health insurance - your Australian Government rebate received \nno.  \
0                                               3557                             
1                                               1141                             
2                                              10610                             
3                                               3913                             
4                                               4356                             

   Private health insurance - your Australian Government rebate received \n$  \
0                                            2662024                           
1                                             630696                           
2                                            6369968                           
3                                            2761608                           
4                                            2632078                           

   Private health insurance rebate \nno.  Private health insurance rebate \n$  \
0                                    440                               185991   
1                                    213                                63883   
2                                   1461                               474942   
3                                    442                               154634   
4                                    552                               175697   

   Excess private health reduction or refund 

In [11]:
# number of unique values for Postcode variable
postcode_df['Postcode'].nunique()

2567

There are 2567 unique values, and this corresponds exactly to the number of rows of this DataFrame, so we're absolutely sure this column contains unique values. This also means that after merging the two tables, there will be only one-to-one matches. We won't have a case where we get multiple rows from one of the datasets matching with only one row of the other one. For instance, postcode 2029 from the ATO DataFrame will have exactly one match in the second Postcode DataFrame.

In [12]:
# left join
merged_df = df.merge(postcode_df, how='left', on='Postcode')
merged_df.shape

(2473, 177)

In [15]:
for col in merged_df.columns:
    print(col)

Postcode
Count taxable income or loss
Average taxable income or loss
Median taxable income or loss
Count salary and wages
Average salary and wages
Median salary and wages
Count net rent
Average net rent
Median net rent
Count total income or loss
Average total income or loss
Median total income or loss
Count total deductions
Average total deductions
Median total deductions
Count total business income
Average total business income
Median total business income
Count  total business expenses
Average total business expenses
Median total business expenses
Count net tax
Average net tax
Median net tax
Count super total accounts balance
Average super total accounts balance
Median super total accounts balance
State/ Territory1
Number of individuals
no.
Taxable income or loss3 
no.
Taxable income or loss3 
$
Tax on taxable income 
no.
Tax on taxable income 
$
Medicare levy 
no.
Medicare levy 
$
Medicare levy surcharge 
no.
Medicare levy surcharge 
$
Total Medicare levy liability 
no.
Total Medica

We got exactly 2473 rows after merging, which is what we expect as we used a left join and there was a one-to-one match on the Postcode column from both original DataFrames. Also, we now have 177 columns, which is the objective of this exercise. But before concluding it, we want to see whether there are any postcodes that didn't match between the two datasets. To do so, we will be looking at one column from the right-hand side DataFrame (the Postcode dataset) and see if there are any missing values.

In [16]:
# total missing values for 'State/Territory1'
merged_df['State/ Territory1'].isna().sum()

4

In [18]:
# print missing postcodes using .iloc()
merged_df.loc[merged_df['State/ Territory1'].isna(), 'Postcode']

631     3010
1494    4462
2072    6068
2332    6758
Name: Postcode, dtype: object

The missing postcodes from the Postcode dataset are 3010, 4462, 6068, and 6758. In a real project, you would have to get in touch with your stakeholders or the data team to see if you are able to get this data.